File: iosApp/Sources/Components/TransportButton.swift
import SwiftUI

/// One cell of the player's transport rows: a flat, full-bleed button that
/// distinguishes a tap from a press-and-hold. Shared by `FrameStepBar` and
/// `PlaybackControlBar` so the two rows stack as one control surface.
///
/// Port of Android's TransportButton (androidApp .../clipdetail/TransportButton.kt).
/// `onTap` fires immediately on press so repeated taps stay responsive;
/// `onHoldActivate` fires once after `holdActivationDelay`; `onHoldTick` repeats
/// afterwards, and a nil tick means a hold is a one-shot gesture.
struct TransportButton: View {
    let text: String
    var accessibilityText: String? = nil
    var onTap: () -> Void = {}
    var holdActivationDelay: Duration = .milliseconds(400)
    var onHoldActivate: () -> Void = {}
    var onHoldTick: (() -> Void)? = nil
    var holdTickPeriod: Duration = .milliseconds(100)
    var onRelease: () -> Void = {}

    @State private var pressed = false
    @State private var holdTask: Task<Void, Never>? = nil

    var body: some View {
        Text(text)
            .font(.body.weight(.semibold))
            .foregroundStyle(Shuttl.text)
            .lineLimit(1)
            .minimumScaleFactor(0.7)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 12)
            .background(pressed ? Shuttl.bgSecondary : Shuttl.bgTertiary)
            .overlay(Rectangle().stroke(Shuttl.borderSecondary, lineWidth: 1))
            .contentShape(Rectangle())
            .accessibilityLabel(accessibilityText ?? text)
            .gesture(
                DragGesture(minimumDistance: 0)
                    .onChanged { _ in
                        guard !pressed else { return }
                        pressed = true
                        onTap()
                        holdTask = Task {
                            try? await Task.sleep(for: holdActivationDelay)
                            guard !Task.isCancelled else { return }
                            onHoldActivate()
                            guard let tick = onHoldTick else { return }
                            while !Task.isCancelled {
                                tick()
                                try? await Task.sleep(for: holdTickPeriod)
                            }
                        }
                    }
                    .onEnded { _ in
                        endPress()
                    }
            )
            .onDisappear { endPress() }
    }

    private func endPress() {
        holdTask?.cancel()
        holdTask = nil
        pressed = false
        onRelease()
    }
}

File: androidApp/src/main/java/com/badmintontracker/android/clipdetail/TransportButton.kt
package com.badmintontracker.android.clipdetail

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.badmintontracker.android.ui.theme.ShuttlTheme
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * One cell of the player's transport rows: a flat, full-bleed button that
 * distinguishes a tap from a press-and-hold. Shared by [FrameStepBar] and
 * [PlaybackControlBar] so the two rows stack as one control surface.
 *
 * @param onTap fires immediately on press, so repeated taps stay responsive.
 * @param onHoldActivate fires once, [holdActivationDelayMs] after the press.
 * @param onHoldTick repeats every [holdTickPeriodMs] after activation. Null
 *   means a hold is a one-shot gesture and no repeat loop is started.
 */
@Composable
internal fun TransportButton(
    text: String,
    modifier: Modifier = Modifier,
    contentDescription: String? = null,
    onTap: () -> Unit = {},
    holdActivationDelayMs: Long = 400L,
    onHoldActivate: () -> Unit = {},
    onHoldTick: (() -> Unit)? = null,
    holdTickPeriodMs: Long = 100L,
    onRelease: () -> Unit = {},
) {
    val scope = rememberCoroutineScope()
    var pressed by remember { mutableStateOf(false) }
    // The gesture detector is started once and outlives recomposition, so it must
    // read the callbacks through state - captured lambdas would keep firing with
    // the skip interval that was current when the button first composed.
    val tap by rememberUpdatedState(onTap)
    val holdActivate by rememberUpdatedState(onHoldActivate)
    val holdTick by rememberUpdatedState(onHoldTick)
    val release by rememberUpdatedState(onRelease)
    val bg = ShuttlTheme.extended.bgTertiary
    val pressedBg = MaterialTheme.colorScheme.surfaceVariant
    val fg = MaterialTheme.colorScheme.onSurface
    val borderColor = MaterialTheme.colorScheme.outline

    Row(
        modifier = modifier
            .background(if (pressed) pressedBg else bg)
            .border(width = 1.dp, color = borderColor)
            .pointerInput(Unit) {
                detectTapGestures(
                    onPress = {
                        pressed = true
                        tap()
                        val holdJob = scope.launch {
                            delay(holdActivationDelayMs)
                            holdActivate()
                            while (isActive) {
                                val tick = holdTick ?: break
                                tick()
                                delay(holdTickPeriodMs)
                            }
                        }
                        try {
                            tryAwaitRelease()
                        } finally {
                            holdJob.cancel()
                            pressed = false
                            release()
                        }
                    },
                )
            }
            .then(
                if (contentDescription != null) {
                    Modifier.semantics { this.contentDescription = contentDescription }
                } else {
                    Modifier
                }
            )
            .padding(horizontal = 16.dp, vertical = 12.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = text,
            color = fg,
            fontWeight = FontWeight.SemiBold,
            textAlign = TextAlign.Center,
            style = MaterialTheme.typography.bodyLarge,
        )
    }
}
